streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java

```diff
@@ -270,22 +233,22 @@ public void flush() {
         }
     }
 
-    private void logValue(final Bytes key, final BufferKey bufferKey, final ContextualRecord value) {
-        final byte[] serializedContextualRecord = value.serialize();
+    private void logValue(final Bytes key, final BufferKey bufferKey, final BufferValue value) {
+        final byte[] serializedBufferValue = value.serialize();
 
         final int sizeOfBufferTime = Long.BYTES;
-        final int sizeOfContextualRecord = serializedContextualRecord.length;
+        final int sizeOfBufferValue = serializedBufferValue.length;
```

Comment:
I know this is not introduced in this PR, but could we consider not allocating the buffer multiple times (here we would allocate three times, each in different levels of the object's `serialize()` function).

Reply:
Nice challenge! I managed to pull this off by extending the serialization logic.